File: internal/bot/command/check_app_scores.go
package command

import (
	"context"
	"fmt"
	"math"
	"strings"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"github.com/miti99/store-scraper-bot-go/internal/api/apple"
	"github.com/miti99/store-scraper-bot-go/internal/api/google"
	"github.com/miti99/store-scraper-bot-go/internal/config"
	"github.com/miti99/store-scraper-bot-go/internal/model"
	"github.com/miti99/store-scraper-bot-go/internal/repository"
	"github.com/miti99/store-scraper-bot-go/internal/util"
)

// /checkappscore — Java CheckAppScoreCommand. Reports score + ratings count.
// Score is rounded to 1 decimal (Java Precision.round(score, 1) parity).
type CheckAppScoresCommand struct {
	cfg           *config.Config
	adminRepo     *repository.AdminRepository
	groupRepo     *repository.GroupRepository
	appleScraper  *apple.AppleScraper
	googleScraper *google.GoogleScraper
}

func NewCheckAppScoresCommand(cfg *config.Config, adminRepo *repository.AdminRepository, groupRepo *repository.GroupRepository, a *apple.AppleScraper, g *google.GoogleScraper) *CheckAppScoresCommand {
	return &CheckAppScoresCommand{cfg: cfg, adminRepo: adminRepo, groupRepo: groupRepo, appleScraper: a, googleScraper: g}
}

func (c *CheckAppScoresCommand) Execute(msg *tgbotapi.Message, sender Sender) {
	if !authorizeGroup(msg.Chat.ID, c.adminRepo, sender) {
		return
	}
	if len(splitArgs(msg.CommandArguments())) != 0 {
		_ = sender.SendMessage(msg.Chat.ID, "Invalid arguments")
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()
	group, err := c.groupRepo.Get(ctx, msg.Chat.ID)
	if err != nil {
		_ = sender.SendMessage(msg.Chat.ID, "Internal server error")
		return
	}

	headers := []string{"AppId", "Score", "Ratings"}
	appleRows := c.appleScoreRows(group.AppleApps)
	googleRows := c.googleScoreRows(group.GoogleApps)

	var sb strings.Builder
	sb.WriteString("<b>Apple Apps</b>\n")
	if len(appleRows) == 0 {
		sb.WriteString("<i>(none)</i>\n")
	} else {
		sb.WriteString(fmt.Sprintf("<pre>%s</pre>\n", util.BuildTable(headers, appleRows)))
	}
	sb.WriteString("\n<b>Google Apps</b>\n")
	if len(googleRows) == 0 {
		sb.WriteString("<i>(none)</i>\n")
	} else {
		sb.WriteString(fmt.Sprintf("<pre>%s</pre>\n", util.BuildTable(headers, googleRows)))
	}
	_ = sender.SendMessage(msg.Chat.ID, sb.String())
}

func (c *CheckAppScoresCommand) appleScoreRows(apps []model.AppInfo) [][]string {
	rows := make([][]string, 0, len(apps))
	for _, a := range apps {
		resp, err := c.appleScraper.GetApp(a.AppID, a.Country)
		if err != nil || resp == nil {
			rows = append(rows, []string{a.AppID, "?", "?"})
			continue
		}
		rows = append(rows, []string{a.AppID, formatScore(resp.Score), fmt.Sprintf("%d", resp.Ratings)})
	}
	return rows
}

func (c *CheckAppScoresCommand) googleScoreRows(apps []model.AppInfo) [][]string {
	rows := make([][]string, 0, len(apps))
	for _, a := range apps {
		resp, err := c.googleScraper.GetApp(a.AppID, a.Country)
		if err != nil || resp == nil {
			rows = append(rows, []string{a.AppID, "?", "?"})
			continue
		}
		rows = append(rows, []string{a.AppID, formatScore(resp.Score), fmt.Sprintf("%d", resp.Ratings)})
	}
	return rows
}

// formatScore rounds to 1 decimal place (Java Precision.round(score, 1)).
func formatScore(score float64) string {
	rounded := math.Round(score*10) / 10
	return fmt.Sprintf("%.1f", rounded)
}

File: src/bot/commands/check-app-scores.js
import * as groupRepo from '../../repository/group-repository.js';
import { buildTable } from '../../util/table.js';
import { authorizeGroup, getCommandArguments, splitArgs } from './command-utils.js';

// /checkappscore — Java CheckAppScoreCommand. Reports score + ratings.
// Score rounded to 1 decimal (Java Precision.round(score, 1) parity).
export function createCheckAppScoresCommand(appleScraper, googleScraper) {
  return async (msg, sender) => {
    if (!(await authorizeGroup(msg.chat.id, sender))) return;
    if (splitArgs(getCommandArguments(msg.text)).length !== 0) {
      await sender.sendMessage(msg.chat.id, 'Invalid arguments');
      return;
    }
    try {
      const group = await groupRepo.getGroup(msg.chat.id);
      const headers = ['AppId', 'Score', 'Ratings'];
      const appleRows = await scoreRowsFor(group.appleApps, appleScraper);
      const googleRows = await scoreRowsFor(group.googleApps, googleScraper);

      const out =
        '<b>Apple Apps</b>\n' +
        renderTable(appleRows, headers) +
        '\n<b>Google Apps</b>\n' +
        renderTable(googleRows, headers);
      await sender.sendMessage(msg.chat.id, out);
    } catch {
      await sender.sendMessage(msg.chat.id, 'Internal server error');
    }
  };
}

async function scoreRowsFor(apps, scraper) {
  const rows = [];
  for (const a of apps) {
    try {
      const resp = await scraper.getApp(a.appId, a.country);
      if (!resp) {
        rows.push([a.appId, '?', '?']);
        continue;
      }
      rows.push([a.appId, formatScore(resp.score), String(resp.ratings ?? 0)]);
    } catch {
      rows.push([a.appId, '?', '?']);
    }
  }
  return rows;
}

function renderTable(rows, headers) {
  if (rows.length === 0) return '<i>(none)</i>\n';
  return `<pre>${buildTable(headers, rows)}</pre>\n`;
}

function formatScore(score) {
  const v = Number(score);
  if (!Number.isFinite(v)) return '?';
  return (Math.round(v * 10) / 10).toFixed(1);
}
